Replace crate collection via crushing with dedicated crate collection logic

reaperrr opened the pull request:
- Allows crates to be collected by non-mobile actors (aircraft, buildings etc.)
- Allows crates to be hidden behind buildings, like in the original games (by not giving or removing `CrateCollector` to/from building in question)
- Allows paratroops being dropped on cells occupied by crates
- Allows mobile actors to enter cells that a crate is currently being paradropped on

For testing, I suggest adding `CrateCollector` to helicopter default in RA, then ingame enable insta-build cheat and test a) placing a building under a dropping crate, b) place a building on a dropped crate, and c) let a transport heli land on a crate. All of these should work.

Fixes #9555.
Closes #14920.

reaperrr committed:
Refactor crate collection logic

Replaced crush-based collecting with dedicated
'collector entered crate-cell' check.

- Allows crates to be collected by non-mobile actors,
  like aircraft, buildings etc.
- Allows crates to be hidden behind buildings,
  like in the original games
- Allows paratroops being dropped on cells with crates
- Allows mobile actors to enter cells that a crate
  is currently being paradropped on

reaperrr committed:
Adapt official mods to crate collect refactor

This also allows buildings in the official mods to collect crates
when a player places a building on a crate.

However, due to balance implications or lack of effect,
invulnerability in RA, stealth in TD and stealth + power-ups in TS
have been disabled.

The following effects would have been filtered as well, but:
- LevelUpCrateAction already filters out actors that
  don't have GainsExperience (hence why no D2k changes)
- DuplicateUnitCrateAction already filters out actors without IPositionable
- GiveMcvCrateAction can only trigger when player has no CYard,
  so no filter needed (no way to place building)

reaperrr committed:
CrateCollector update rule

GraionDilach commented:
Fixes #9555.

reaperrr commented:
Updated.

reaperrr commented:
Rebased and updated update rule.

GraionDilach commented:
Why are you adding the upgrade rule directly to a folder? Am I unaware of something?

reaperrr commented:
I was under the assumption the update path date won't change anymore, now that we've branched off the playtest, so unless this PR will sit unmerged until after the Next+1 branch-off, it would belong into that folder?

pchote commented:
The convention is to have the folders named after the final release, with bleed-only rules not having a folder until we do the prep branchoff.

reaperrr commented:
Updated, moved rule out of folder.

Smittytron commented:
> For testing, I suggest adding CrateCollector to helicopter and building defaults in RA, then ingame enable inst-build cheat and test a) placing a building under a dropping crate, b) place a building on a dropped crate, and c) let a transport heli land on a crate. All of these should work.

Did this, and everything but b) worked. Single cell structures would catch the crates fine, but multi cell structures did not collect the crates. I also had a crate drop directly onto an ore mine but I'm not sure if that was already an issue.

pchote commented:
Needs a rebase.

pchote commented on OpenRA.Mods.Common/OpenRA.Mods.Common.csproj:
```diff
@@ -911,6 +912,7 @@
     <Compile Include="UpdateRules\Rules\20180307\AddShakeToBridge.cs" />
     <Compile Include="UpdateRules\Rules\20180307\IgnoreAbstractActors.cs" />
     <Compile Include="UpdateRules\Rules\20180307\RemoveCanUndeployFromGrantConditionOnDeploy.cs" />
+    <Compile Include="UpdateRules\Rules\AddCrateCollector.cs" />
```
This will need to move too.

pchote commented on OpenRA.Mods.Common/Traits/CrateCollector.cs:
```diff
@@ -0,0 +1,37 @@
+	public class CrateCollectorInfo : ITraitInfo
+	{
+		[Desc("E.g. crates. Leave empty if actor is supposed to be able to collect everything.")]
+		public readonly HashSet<string> Collects = new HashSet<string>();
```
Lets use a `BitSet` here.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/AddCrateCollector.cs:
```diff
@@ -0,0 +1,141 @@
+
+		public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNode actorNode)
+		{
+			var crate = actorNode.LastChildMatching("Crate");
```
Can we please loop over ChildrenMatching? (ref the pr I reviewed earlier tonight).

reaperrr commented:
Rebased and updated.

> Single cell structures would catch the crates fine, but multi cell structures did not collect the crates. I also had a crate drop directly onto an ore mine but I'm not sure if that was already an issue.

Identified and fixed both issues.

Pushed a testcase here for easier reviewing (RA mod, makes buildings crate collectors and increases spawn rate): https://github.com/reaperrr/OpenRA/tree/crate-collect-test

abcdefg30 reviewed (dismissed):
You can now construct buildings on crates (even without the testcase).
![grafik](https://user-images.githubusercontent.com/7704140/46232060-3a7eaf80-c36e-11e8-8af6-b9e5d9bbe4af.png)
Imho we should either prohibit this or (the better option I think) enable crate collection on buildings by default.

CH4Code commented:
Veteran buildings, spawning a cluster of teslas instead of one(multiply crate)? Exploding a placed structure, last-minute-IC on deployed mcv, all the fun stuff :)

MustaphaTR commented:
Needs rebase.

Also i think i'll need this for Sole Survivor mod (i plan to add the aircraft, even tho it wasn't in the original). So would like to see this resolved soon. I can test myself after a rebase.

reaperrr commented:
> Imho we should either prohibit this or (the better option I think) enable crate collection on buildings by default.

The former would partially defeat the purpose of this PR, namely making crates not block space.
To avoid that as well as the balance implications of the latter, my proposal would be to:
- add an `ExcludeCollectorTypes` filter to `CrateAction`, then 
- in the shipping mods, adapt the yaml side to make them have no effect when collected by buildings (anything else would be too hard/take too long to get balanced right within the scope of this PR).

MustaphaTR commented:
> add an ExcludeCollectorTypes filter to CrateAction, then

Afaik we already have that.

reaperrr commented:
> Afaik we already have that.

Only for actors, meaning you'd have to list every single actor you want to exclude.
Excluding collector classes is much more convenient to use.

reaperrr commented:
Updated and rebased.

I changed my mind and only disabled potentially balance-breaking crate-effects for buildings, namely invul in RA, stealth in TD, and stealth in TS. I also disabled power-ups in TS, though they would simply have no effect anyway, as there are no traits on buildings that could get enabled by that condition.

Some of the other potentially problematic CrateActions actually already filter out buildings in our shipping mods (see desc of 3rd commit).

Note: I intentionally left explosions enabled, otherwise collecting crates via buildings would have a too high chance of a positive effect compared to units and everyone would start to use buildings to collect crates within the base radius whenever possible.

MustaphaTR commented:
I tested, but it looks a bit wierd that when a building collects a crate, anim/cash ticks plays at building center, units from GiveUnitCrateAction are spawned around top-left of the building. Imo it would be better if we change these to happen at the crate's position.

Also i couldn't get aircraft to collect the crates, but not sure if i'm doing something wrong with the yaml or is it actually broken.

pchote commented:
> Also i couldn't get helicopters to collect the crates, but not sure if i'm doing something wrong with the yaml or is it actually broken.

Helicopters still don't insert themselves into the actormap (we have an issue but I can't find it), so this indeed won't work. Not a problem with this PR though.

Needs a rebase.

pchote commented:
Note from discussions with @teinarss in IRC that this and a similar change to mines are prereqs for the Locomotor pathability cache PR.

reaperrr commented:
> Helicopters still don't insert themselves into the actormap (we have an issue but I can't find it), so this indeed won't work. Not a problem with this PR though.

I'm working around that by checking a radius in addition to the ActorMap, so it should work if enabled in yaml.

However, I've not enabled that for any helis in RA/TD, as their speed might make them 'too good' at that and affect balance. To test it, you'd have to add CrateCollector to the heli default (or directly to TRAN) in either mod.

abcdefg30 commented:
I would consider transports (the only aircraft that can land to collect crates in the default mods) being able to collect crates a bug fix.

reaperrr commented:
> I would consider transports (the only aircraft that can land to collect crates in the default mods) being able to collect crates a bug fix.

Ok, convinced :)
Updated & rebased.

abcdefg30 commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -91,47 +101,41 @@ public Crate(ActorInitializer init, CrateInfo info)
-				.Where(a => a != self);
+			// HACK: Currently needed to find aircraft actors.
+			// TODO: Remove this once GetActorsAt supports aircraft.
+			sameCellActors.Concat(self.World.FindActorsInCircle(self.CenterPosition, new WDist(724))
```
Couldn't we drop the `GetActorsAt` call entirely for now if we are using this afterwards anyway?

reaperrr replied:
That's what I thought as well, until I ran into everyone's favorite issue of larger buildings' CenterPosition being potentially too far away.
I guess I could use `LargestActorRadius` to fix that, in terms of performance it's probably not much of a difference anyway.

pchote replied:
Throwing out an idea: if we went this route we could expose the range to yaml, providing proximity collection for some usecases (e.g. proximity mines?)

abcdefg30 reviewed (commented):
Needs a rebase again...

pchote commented:
This only conflicts with the update rules, so shouldn't block reviewing.

pchote commented on OpenRA.Mods.Common/Traits/Crates/CrateAction.cs:
```diff
@@ -65,6 +69,10 @@ public int GetSelectionSharesOuter(Actor collector)
 				return 0;
 
+			var crateCollectorTraitInfo = collector.Info.TraitInfo<CrateCollectorInfo>();
+			if (!crateCollectorTraitInfo.All && crateCollectorTraitInfo.CollectorTypes.Overlaps(Info.ExcludeCollectorTypes))
```
```csharp
 if (!crateCollectorTraitInfo.CollectorTypes.IsEmpty && crateCollectorTraitInfo.CollectorTypes.Overlaps(Info.ExcludeCollectorTypes)) 
```

pchote commented on OpenRA.Mods.Common/Traits/CrateCollector.cs:
```diff
@@ -0,0 +1,38 @@
+		[Desc("Define collector type(s) checked by Crate and CrateAction for validity. Leave empty if actor is supposed to be able to collect any crate.")]
+		public readonly BitSet<CollectorType> CollectorTypes = default(BitSet<CollectorType>);
+
+		public bool All { get { return CollectorTypes == default(BitSet<CollectorType>); } }
```
This is already provided by `BitSet<T>.IsEmpty`, so can be removed.

pchote commented on OpenRA.Mods.Common/Traits/CrateCollector.cs:
```diff
@@ -0,0 +1,38 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("This actor can collect crates.")]
+	public class CrateCollectorInfo : ITraitInfo
```
Are there any gotchas arguing against making this conditional?

pchote commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -26,15 +29,18 @@ class CrateInfo : ITraitInfo, IPositionableInfo, Requires<RenderSpritesInfo>
+
+		[Desc("Set this to 'true' to allow crates to spawn on cells occupied by other actors. " +
+			"Note that this might make the crate unreachable if spawned on immobile, unremovable actors.")]
+		public readonly bool IgnoreActorsOnSpawning = false;
```
What is the usecase for setting this `true`?

reaperrr replied:
Modders could want crates to be paradroppable on buildings. But I guess I should either drop this or add a valid collector check to avoid that 'unreachable' issue.

pchote commented:
Bumping to Next + 1.

reaperrr commented:
Sorry for the lack of progress here, but I want to focus on the aircraft and reservable/resupply refactors (and smaller PRs) first, otherwise I fear neither might be ready by the time we branch off prep+1.

reaperrr commented:
Not sure how to go about this right now, and as things stand this would stay stale for months. Closing for now.
